### DISCUSSION TITLE
What is the correct command to show the current open buffer diff against the latest git commit (HEAD) in a new tab? (no plugins)

### DISCUSSION THREAD

**Discussion by charbelnicolas** [neovim/neovim#31237] [Q&A]
Hello, I've been trying to come up with a command that will show a diff of the current open buffer against the latest git commit (HEAD) in a new tab but I just can't seem to make it work. I don't want to use a plugin like vim-fugitive. Any idea of how to make this work?

Thank you

**Comment by telemachus** [comment]
Check out this answer on Stackexchange: https://vi.stackexchange.com/a/28815. I had to tweak the first command, but the following worked for me.

```vimscript
:tabnew | r! git show HEAD^:$(git rev-parse --show-prefix)#1
:vert sb 1 | windo diffthis
```

From a comment on that post, the following also works, and it doesn't rely on the specific tab number in the first version.

```vimscript
:tabedit % | diffthis | vertical new | diffthis | read !git show HEAD^:#
```

**Reply by charbelnicolas** [reply]
> Check out this answer on Stackexchange: https://vi.stackexchange.com/a/28815. I had to tweak the first command, but the following worked for me.
> 
> ```viml
> :tabnew | r! git show HEAD^:$(git rev-parse --show-prefix)#1
> :vert sb 1 | windo diffthis
> ```
> 
> From a comment on that post, the following also works, and it doesn't rely on the specific tab number in the first version.
> 
> ```viml
> :tabedit % | diffthis | vertical new | diffthis | read !git show HEAD^:#
> ```

Hi @telemachus, I tried using the first option and it didn't work but the second one did. The only problem is that it doesn't show tree-sitter highlights and is also shown as an unsaved buffer/file.

I was thinking that it was possible to get a diff just like doing it on the terminal with `git difftool`. It even uses neovim to do it... so it is kind of awkward that there is no easy direct way to show the diff with a command inside neovim.

Thank you!

**Comment by kkkk0524** [comment]
I wrote a script implementing similar functionality: using git via diff command in nvim, applied to buffers in neovim. By default, it compares the local file with HEAD (':Df'), and HEAD with HEAD~ (':Df 1 2'). However, I'm on Windows and the code may need adjustments.
```lua
function _G.custom_diff(far,near)
  if (near > far) then
    far, near = near, far
  end 

  local temp_path = vim.env.TEMP or vim.env.TMP
  temp_path = temp_path:gsub("\\", "/")

  local file_path = vim.fn.expand('%:p')
  local file_name = vim.fn.expand('%:t')
  local open_far, open_near, far_path, near_path
  if near == 0 then
    far = far - 1
    local git_root = vim.fn.systemlist('git rev-parse --show-toplevel')[1]
    git_root = git_root:gsub('\\', '/'):gsub('/$', '')
    file_path = file_path:gsub('\\', '/')
    -- 提取相对路径
    local relative_path = file_path:sub(#git_root + 2)
    local far_head = (far >= 1) and ("HEAD~" .. far) or "HEAD"

    local far_head_prefix = (far >= 1) and ("HEAD~" .. far .. "_") or "HEAD_"
    far_path = temp_path .. "/" .. far_head_prefix .. file_name
    near_path = temp_path .. "/" .. file_name

    open_far = "!git show " .. far_head .. ":" .. relative_path .. " > " .. far_path
    open_near = "lua vim.fn.writefile(vim.fn.readfile(\""  .. file_path .. "\"), \"" .. near_path .. "\")"

  else
    far = far - 1
    near = near - 1

    local git_root = vim.fn.systemlist('git rev-parse --show-toplevel')[1]
    git_root = git_root:gsub('\\', '/'):gsub('/$', '')
    file_path = file_path:gsub('\\', '/')
    local relative_path = file_path:sub(#git_root + 2)


    local far_head = (far >= 1) and ("HEAD~" .. far) or "HEAD"
    local near_head = (near > 0) and ("HEAD~" .. near) or "HEAD"

    local far_head_prefix = (far >= 1) and ("HEAD~" .. far .. "_") or "HEAD_"
    local near_head_prefix = (near > 0) and ("HEAD~" .. near .. "_") or "HEAD_"
    far_path = temp_path .. "/" .. far_head_prefix .. file_name
    near_path = temp_path .. "/" .. near_head_prefix .. file_name

    open_far = "!git show " .. far_head .. ":" .. relative_path .. " > " .. far_path
    open_near = "!git show " .. near_head .. ":" .. relative_path .. " > " .. near_path
  end
  vim.cmd(open_far)
  vim.cmd(open_near)

  local open_diff = "edit! " .. far_path .. " | vert diffsplit " .. near_path
  vim.cmd(open_diff)
end

vim.api.nvim_create_user_command('Df', function(opts)
  local args = opts.fargs
  local far = tonumber(args[1]) or 1
  local near = tonumber(args[2]) or 0
  _G.custom_diff(far, near)
end, {
    nargs = '*',
    complete = function() return {"1", "2", "3"} end,
    desc = "Git diff"
  })
```

**Reply by charbelnicolas** [reply]
Thank you, I'll try it when I get back home, I'm on linux ;)
